universal-login-commons/src/core/utils/InitializationHandler.ts

```diff
@@ -0,0 +1,68 @@
+type State<T, B> = {
+  kind: 'None';
+} | {
+  kind: 'Initializing';
+  value: T | Promise<T>;
+} | {
+  kind: 'Initialized';
+  value: T | Promise<T>;
+} | {
+  kind: 'Finalizing';
+  value: B | Promise<B>;
+} | {
+  kind: 'Finalized';
+  value: B | Promise<B>;
+};
+
+export class InitializationHandler<T, B> {
+  private state: State<T, B> = {kind: 'None'};
+
+  constructor(
+    private _initialize: () => (T | Promise<T>),
+    private _finalize: () => B | Promise<B>,
+  ) {}
+
+  initialize(): T | Promise<T> {
+    if (this.state.kind === 'Finalizing') {
+      throw Error('Cannot initialize during finalizing');
+    }
+    if (this.state.kind === 'Finalized' || this.state.kind === 'None') {
+      const value = this._initialize();
+      if (isPromise(value)) {
+        this.state = {
+          kind: 'Initializing',
+          value,
+        };
+        (this.state.value as Promise<T>).then(() => {
+          this.state = {kind: 'Initialized', value};
+        });
+      } else {
+        this.state = {kind: 'Initialized', value};
+      }
+    }
+    return this.state.value;
+  }
+
+  finalize(): B | Promise<B> | undefined {
+    if (this.state.kind === 'None') {
+      return;
+    }
+    if (this.state.kind === 'Initializing') {
+      throw Error('Cannot finalize during initializing');
+    }
+    if (this.state.kind === 'Initialized') {
+      const value = this._finalize();
+      if (isPromise(value)) {
+        this.state = {kind: 'Finalizing', value};
+        (value as Promise<B>).then(() => {
+          this.state = {kind: 'Finalized', value};
+        });
+      } else {
+        this.state = {kind: 'Finalized', value};
+      }
+    }
+    return this.state.value;
+  }
+}
+
+const isPromise = (value: unknown) => value && !!(value as any).then;
```

Comment:
```suggestion
const isPromise = (value: unknown): value is Promise<any> => value && !!(value as any).then;
```
This tells typescript that this function check for value to be a promise. It will also update the types if you use this in an if statement.